src/clients/repos.js

```diff
@@ -0,0 +1,122 @@
+
+const COMMIT_MESSAGE = 'Created using Popcode: https://popcode.org';
+const MASTER_BRANCH = 'master';
+
+// Sometimes there are race conditions with creating multiple files in separate
+// API requests
+function repoPerformWithRetries(perform) {
+  return performWithRetries(perform, ['Not Found']);
+}
+
+async function createHtmlFile(github, project, userName, repoName) {
+  if (trim(project.sources.html)) {
+    return github.getRepo(userName, repoName).
+      writeFile(
+        MASTER_BRANCH,
+        'raw_index.html',
+        project.sources.html,
+        COMMIT_MESSAGE,
+        {},
+      );
+  }
+  return Promise.resolve();
+}
+
+async function createCssFile(github, project, userName, repoName) {
+  if (trim(project.sources.css)) {
+    return github.getRepo(userName, repoName).
+      writeFile(
+        MASTER_BRANCH,
+        'styles.css',
+        project.sources.css,
+        COMMIT_MESSAGE,
+        {},
+      );
+  }
+  return Promise.resolve();
+}
+
+async function createJsFile(github, project, userName, repoName) {
+  if (trim(project.sources.javascript)) {
+    return github.getRepo(userName, repoName).writeFile(
+      MASTER_BRANCH, 'script.js', project.sources.js, COMMIT_MESSAGE, {});
+  }
+  return Promise.resolve();
+}
+
+async function createPreviewFile(github, project, userName, repoName) {
+  if (trim(project.sources.html) ||
+    trim(project.sources.css) ||
+    trim(project.sources.javascript)) {
+    const preview = generatePreview(
+      project,
+      {
+        targetBaseTop: false,
+        propagateErrorsToParent: false,
+        breakLoops: false,
+        nonBlockingAlertsAndPrompts: false,
+      },
+    );
+    return github.getRepo(userName, repoName).
+      writeFile(MASTER_BRANCH, 'index.html', preview, COMMIT_MESSAGE, {});
+  }
+  return Promise.resolve();
+}
+
+async function createFiles(github, project, userName, repoName) {
+  await repoPerformWithRetries(
+    () => createHtmlFile(github, project, userName, repoName),
+  );
+  await repoPerformWithRetries(
+    () => createCssFile(github, project, userName, repoName),
+  );
+  await repoPerformWithRetries(
+    () => createJsFile(github, project, userName, repoName),
+  );
+  await repoPerformWithRetries(
+    () => createPreviewFile(github, project, userName, repoName),
+  );
+}
+
+async function createBranch(github, userName, repoName, oldBranch, newBranch) {
+  return github.getRepo(userName, repoName).
+    createBranch(oldBranch, newBranch);
+}
+
+async function updateRepoDescription(github, userName, repoName) {
+  const url = `https://${userName}.github.io/${repoName}`;
+  return github.getRepo(userName, repoName).
+    updateRepository({
+      name: repoName,
+      description: url,
+      homepage: url,
+    });
+}
+
+export async function createRepoFromProject(project, user) {
+  const github = gitHub.clientForUser(user);
+  const REPO_NAME = `popcode-repo-${Date.now()}`;
```

Comment:
I took the repo name from the title, naively replacing spaces with '-' for now - I'm sure there are all sorts of other invalid github repository names though, how do we want to handle that?

And I left the timestamp in for now, in the future we'll somehow want to link repositories and send new exports are updates to the existing repo but at this point I don't want to touch that